contracts/mesh-provider/src/ibc.rs

```diff
@@ -213,33 +221,128 @@ pub fn ack_list_validators(
     for val in res.validators {
         VALIDATORS.save(deps.storage, &val, &Validator::new())?;
     }
-    Ok(IbcBasicResponse::new())
+    RETRIES.update(deps.storage, |mut r| -> Result<crate::state::RetryState, StdError> {
+        r.list_validators_reset();
+        Ok(r)
+    })?;
+    Ok(IbcBasicResponse::new()
+        .add_attribute("action", "ack list_validators"))
 }
 
-pub fn fail_list_validators(_deps: DepsMut) -> Result<IbcBasicResponse, ContractError> {
-    // TODO: send another ListValidators message
-    unimplemented!();
+pub fn fail_list_validators(deps: DepsMut, env: Env) -> Result<IbcBasicResponse, ContractError> {
+    // check if we should retry
+    let mut retries = RETRIES.load(deps.storage)?;
+    if !retries.list_validators_should_retry() {
+        retries.list_validators_reset();
+        RETRIES.save(deps.storage, &retries)?;
+        return Ok(IbcBasicResponse::new().add_event(Event::new("list_validators_fail")))
+    }
+    RETRIES.save(deps.storage, &retries)?;
+
+    // retry
+    let channel_id = CHANNEL.load(deps.storage)?;
+    let packet = ProviderMsg::ListValidators {};
+    let msg = IbcMsg::SendPacket {
+        channel_id: channel_id,
+        data: to_binary(&packet)?,
+        timeout: build_timeout(&env),
+    };
+    Ok(IbcBasicResponse::new()
+        .add_event(Event::new("list_validators_retry"))
+        .add_message(msg))
+}
+
+fn ack_stake(deps: DepsMut, staker: String, validator: String, amount: Uint128) -> Result<IbcBasicResponse, ContractError> {
+    let staker = deps.api.addr_validate(&staker)?;
+
+    if amount.is_zero() {
+        return Ok(IbcBasicResponse::new().add_event(Event::new("ack_stake_zero_amount")));
+    }
+
+    let mut val = VALIDATORS
+        .may_load(deps.storage, &validator)?
+        .ok_or_else(|| ContractError::UnknownValidator(validator.clone()))?;
+    let mut stake = STAKED.load(deps.storage, (&staker, &validator))?;
+    stake.stake_validator(&mut val, amount);
+    STAKED.save(deps.storage, (&staker, &validator), &stake)?;
+    VALIDATORS.save(deps.storage, &validator, &val)?;
+    RETRIES.update(deps.storage, |mut r| -> Result<crate::state::RetryState, StdError> {
+        r.stake_reset();
+        Ok(r)
+    })?;
+
+    Ok(IbcBasicResponse::new()
+        .add_event(Event::new("ack_stake")))
 }
 
 pub fn fail_stake(
-    _deps: DepsMut,
-    // _staker is the staker Addr, not used by consumer
-    _staker: String,
-    _validator: String,
-    _amount: Uint128,
+    deps: DepsMut,
+    env: Env,
+    staker: String,
+    validator: String,
+    amount: Uint128,
 ) -> Result<IbcBasicResponse, ContractError> {
-    // TODO: release the bonded stake, adjust numer
-    unimplemented!();
+    let mut retries = RETRIES.load(deps.storage)?;
```

Comment:
Remove all retries logic here. Only on list_validators.

What we **do** need to do is actually return the stake to the user. We never updated our internal stake/validator book-keeping (that is done on `ack_stake`), but we did receive some bonded tokens from the lockup contract. We should tell that contract that this user will not be bonding this stake and can release it there. Otherwise, those tokens remain locked forever in failure mode.

(Look at logic called on successful claim, after unbonding, to see how to release it)

Reply:
Sorry for my absence, I've been onboarding to a new job. Is it safe to assume we should be returning back 
`coin(amount.u128(), config.rewards_ibc_denom)`?